### ISSUE TITLE
Growing download latency of a CPU-heavy spider

### ISSUE THREAD

**Issue by lopuhin** [scrapy/scrapy#3197]
The issue manifests itself as a growing latency when the spider is relatively CPU-intensive and is sending a lot of requests. Here is an example python 3 spider, based on scrapy bench spider:

```
from urllib.parse import urlencode
import time

import scrapy
from scrapy.linkextractor import LinkExtractor
import logging


class _BenchSpider(scrapy.Spider):
    """A spider that follows all links"""
    name = 'follow'
    total = 10000
    show = 20
    baseurl = 'http://localhost:8998'
    link_extractor = LinkExtractor()
    custom_settings = {
        'LOG_LEVEL': 'INFO',
        'LOGSTATS_INTERVAL': 1,

        'CONCURRENT_REQUEST': 32,  # changed
        'CONCURRENT_REQUEST_PER_DOMAIN': 32,  # changed
    }

    def start_requests(self):
        self.t0 = time.time()
        qargs = {'total': self.total, 'show': self.show}
        url = '{}?{}'.format(self.baseurl, urlencode(qargs, doseq=1))
        return [scrapy.Request(url, dont_filter=True)]

    def parse(self, response):
        # add latency reporting
        logging.info('latency {:.2f} s after {:.0f} s'.format(
            response.meta['download_latency'], time.time() - self.t0))
        for link in self.link_extractor.extract_links(response):
            for i in range(5):  # added CPU work extracting items
                time.sleep(0.05)
                yield {'step': i}
            yield scrapy.Request(link.url, callback=self.parse)
```

Run server in one window: ``python -m scrapy.utils.benchserver``
Run spider: ``scrapy runspider cpu_spider.py``
Ovserve the output:
```
2018-04-02 19:29:50 [scrapy.core.engine] INFO: Spider opened
2018-04-02 19:29:50 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2018-04-02 19:29:50 [root] INFO: latency 0.01 s after 0 s
2018-04-02 19:29:51 [root] INFO: latency 0.11 s after 1 s
2018-04-02 19:29:51 [scrapy.extensions.logstats] INFO: Crawled 3 pages (at 180 pages/min), scraped 17 items (at 1020 items/min)
...
2018-04-02 19:30:02 [root] INFO: latency 0.94 s after 12 s
2018-04-02 19:30:02 [root] INFO: latency 0.94 s after 12 s
2018-04-02 19:30:03 [scrapy.extensions.logstats] INFO: Crawled 34 pages (at 360 pages/min), scraped 239 items (at 2340 items/min)
...
2018-04-02 19:30:57 [root] INFO: latency 8.41 s after 67 s
2018-04-02 19:31:02 [scrapy.extensions.logstats] INFO: Crawled 101 pages (at 480 pages/min), scraped 1318 items (at 5940 items/min)
...
2018-04-02 19:31:56 [root] INFO: latency 12.77 s after 126 s
2018-04-02 19:32:01 [scrapy.extensions.logstats] INFO: Crawled 141 pages (at 480 pages/min), scraped 2410 items (at 8280 items/min)
...
2018-04-02 19:34:36 [root] INFO: latency 20.37 s after 286 s
2018-04-02 19:34:47 [scrapy.extensions.logstats] INFO: Crawled 213 pages (at 480 pages/min), scraped 5470 items (at 12840 items/min)
...
2018-04-02 19:37:02 [root] INFO: latency 25.34 s after 431 s
2018-04-02 19:37:08 [scrapy.extensions.logstats] INFO: Crawled 261 pages (at 480 pages/min), scraped 8103 items (at 16260 items/min)
...
2018-04-02 19:47:57 [root] INFO: latency 40.99 s after 1087 s
2018-04-02 19:48:11 [scrapy.extensions.logstats] INFO: Crawled 413 pages (at 480 pages/min), scraped 20446 items (at 23820 items/min)
2018-04-02 19:48:32 [scrapy.extensions.logstats] INFO: Crawled 413 pages (at 0 pages/min), scraped 20835 items (at 23340 items/min)
```

So it seems that download latency (``response.meta['download_latency']``) is growing linearly, while the actual time the server spends on generating the response is very low.

To summarize, I expected download latency to remain reasonable and not to grow, since the spider is taking constant time to process items.

**Comment by lopuhin** [comment]
The download latency stabilized at around 42-43 seconds now and does not grow any more. For comparison, the spider needs 5 seconds of CPU time for each parse method call (0.05 * 5 * 20 links), although it never blocks the event loop for more than 0.05 seconds.

**Comment by lopuhin** [comment]
Hm, probably this is an expected issue but I'm not sure how to solve it... a simplified Twisted event loop looks like this:
```
                while True:
                    self.runUntilCurrent()
                    self.doIteration()
```
where ``runUntilCurrent`` runs all pending calls, and ``doIteration`` runs I/O. The problem is, all pending calls are run before I/O code is run, so if we have a lot of requests blocking for a small time, then the total time I/O is blocked for is added up across all tasks:
```
2018-04-02 22:15:11 [root] INFO: runUntilCurrent
2018-04-02 22:15:11 [root] INFO: latency 4.22 s after 23 s
2018-04-02 22:15:11 [root] INFO: latency 4.22 s after 23 s
2018-04-02 22:15:12 [root] INFO: latency 4.22 s after 23 s
2018-04-02 22:15:12 [root] INFO: latency 4.22 s after 23 s
2018-04-02 22:15:12 [root] INFO: latency 4.22 s after 23 s
2018-04-02 22:15:12 [root] INFO: latency 2.98 s after 23 s
2018-04-02 22:15:13 [root] INFO: runUntilCurrent done
2018-04-02 22:15:13 [root] INFO: doIteration start
2018-04-02 22:15:14 [root] INFO: doIteration done
```

**Comment by lopuhin** [comment]
If this issue can not be avoided (it seems to me it would be very hard to do so), it would be nice if we could monitor the time between I/O "rounds" (calls to ``doIteration``), and log a warning to the user if the delay between this rounds is larger than some value (several seconds? or more? let it be configurable?), and also update the docs for ``download_latency``.

**Comment by nctl144** [comment]
Do you think this can be added to the gsoc project over the summer? I think it fits into the project category.

**Comment by lopuhin** [comment]
Yes, I think this fits the topic of performance GSoC well, even if solution will only involve better monitoring and documentation.

**Comment by zd-project** [comment]
Yes it looks like a convoy effect to me. The scheduling algorithm prioritizes CPU-intensive operations over I/O operations, which starved I/O operations. I'm wondering if Twisted provides an interface to manipulate scheduling scheme or having alternative algorithm like round-robin under the hood. I'm wondering if it's feasible to wrap Twisted event loop with another buffer to balance operation loads. If not, a warning would be the easiest to notify such problem.

**Comment by zd-project** [comment]
This post from 2011 pointed out that Twisted may only support FCFS scheduling under the hood, [twisted reactor: priority and insights
](https://stackoverflow.com/questions/5159259/twisted-reactor-priority-and-insights). I didn't found on Google whether they've tried to implement to implement alternative scheduling algorithm but I highly doubt they will. 

Instead maybe we can exploit Twisted Deferred interface a bit more to forcefully delay CPU-intensive operations? [Scheduling tasks for the future](http://twistedmatrix.com/documents/13.2.0/core/howto/time.html) I've seen we used this interface already in the `utils/defer.py`, can we build anything from that?

**Comment by lopuhin** [comment]
> Instead maybe we can exploit Twisted Deferred interface a bit more to forcefully delay CPU-intensive operations.

Sorry, I don't understand how simple Deferred will solve the problem, it seems to me that all deferreds will need to be processed in the same way anyway.

Re threads: I think using ``deferToThread`` http://twistedmatrix.com/documents/current/api/twisted.internet.threads.deferToThread.html will help here indeed. This requires some work from the user: first knowing that there is a problem with I/O blocking for a long time (scrapy can help here), and also identifying which operation is causing most slowdown - often this is not obvious and profiling is needed.

**Comment by zd-project** [comment]
Originally I was thinking of capping maximum amount of request per second to ensure fair scheduing, but I guess deferToThread may be a better solution. I think I need to further study on Twisted to understand architecture. I came across this stackoverflow question [is Scrapy single-threaded or multi-threaded?
](https://stackoverflow.com/questions/24761074/is-scrapy-single-threaded-or-multi-threaded), which came out at 2014. But it looks to me scrapy is still single threaded and asynchronous by design, as manifested in `utils/defer.py`? By introducing "deferToThread" we're making modifications to the core of Scrapy architecture here, I'm wondering if this conforms to the design principle of Scrapy. Also it feels like Twisted still prefers single-threaded event loop by its design. See [Using Threads in Twisted](http://twistedmatrix.com/documents/13.2.0/core/howto/threading.html) and [python twisted multithreaded server
](https://stackoverflow.com/questions/19259139/python-twisted-multithreaded-server). Do you have any comments on this?

**Comment by lopuhin** [comment]
@zd-project yes, scrapy and twisted a single-threaded by design and there are no plans to change that. I meant that scrapy user can already put expensive operations into deferToThread and avoid this issue, see for example this answer: https://stackoverflow.com/a/34140412/217088

**Comment by zd-project** [comment]
Interesting. I think the suggestion made for a custom pipeline is enlightening. Do you think it's reasonable to do some similar work with spider? For example, to provide a spider interface that supports low latency download?

**Comment by lopuhin** [comment]
@zd-project I think the same approach works in the spider as well, you can return a deferred from the parse method.

**Comment by zd-project** [comment]
So do we just giving out suggestions in the doc and let users to decide what to do? Or do you suggest we provide a wrapper?

**Comment by zd-project** [comment]
I agree that it’s hard to tell what the event loop would do due to its asynchronous behavior. Maybe we can sample the I/O operations and issue warnings based off that? I can help look into this if needed. @nctl144 do you have any ideas so far?
